File: ComPDFKit for Windows/Examples/Samples/BookmarkTest/VB/BookmarkTest.vb
Imports ComPDFKit.PDFDocument
Imports System
Imports System.IO

Module BookmarkTest
    Private parentPath = Path.GetDirectoryName(Path.GetDirectoryName(Path.GetDirectoryName(System.IO.Directory.GetCurrentDirectory())))
    Private outputPath As String = Path.Combine(parentPath, "Output", "VB")

    Sub Main()
        ' Preparation work
        Console.WriteLine("Running Bookmark test sample…" + Environment.NewLine)
        SDKLicenseHelper.LicenseVerify()
        If Not Directory.Exists(outputPath) Then
            Directory.CreateDirectory(outputPath)
        End If
        Dim document As CPDFDocument = CPDFDocument.InitWithFilePath("ThreeBookmark.pdf")

#Region "Sample 1: Access bookmark"

        If AccessBookmark(document) Then
            Console.WriteLine("Check bookmark list done.")
        Else
            Console.WriteLine("Check bookmark list failed.")
            Console.WriteLine("--------------------")
            Console.WriteLine("Stop.")
            Console.WriteLine("--------------------")
            Console.ReadLine()
            Return
        End If
        document.Release()
        Console.WriteLine("--------------------")

#End Region

#Region "Sample 2: Create bookmark"

        document = CPDFDocument.InitWithFilePath("ThreeBookmark.pdf")

        If CreateBookmark(document) Then
            Console.WriteLine("Add bookmark done.")
        Else
            Console.WriteLine("Add bookmark failed.")
        End If
        document.Release()
        Console.WriteLine("--------------------")

#End Region

#Region "Sample 3: Remove bookmark"

        document = CPDFDocument.InitWithFilePath("ThreeBookmark.pdf")

        If RemoveBookmark(document) Then
            Console.WriteLine("Remove bookmark done.")
        Else
            Console.WriteLine("Remove bookmark failed.")
        End If

        document.Release()
        Console.WriteLine("--------------------")

#End Region

        Console.WriteLine("Done!")
        Console.WriteLine("--------------------")
        Console.ReadLine()



    End Sub

    ''' <summary>
    ''' Access bookmark
    ''' </summary> 
    Private Function AccessBookmark(document As CPDFDocument) As Boolean
        Dim bookmarkList As List(Of CPDFBookmark) = document.GetBookmarkList()

        ' Check if there are 3 bookmarks in the document
        If bookmarkList.Count = 3 Then
            Console.WriteLine("Access bookmark list done.")
        Else
            Return False
        End If

        ' Check the title of the bookmark for page index 0
        If document.BookmarkForPageIndex(0).Title = "Bookmark1" Then
            Console.WriteLine("Access bookmark for a page done.")
        Else
            Return False
        End If

        Return True
    End Function

    ''' <summary>
    ''' Create bookmark
    ''' </summary> 
    Private Function CreateBookmark(document As CPDFDocument) As Boolean
        Dim bookmarkCount As Integer = document.GetBookmarkList().Count

        Dim bookmark As New CPDFBookmark()
        bookmark.Title = "new bookmark"
        bookmark.PageIndex = 4

        ' Add the new bookmark to the document
        document.AddBookmark(bookmark)

        ' Check if the number of bookmarks has increased by 1
        If document.GetBookmarkList().Count - bookmarkCount = 1 Then
            Console.WriteLine("Add bookmark in page {0}. ", bookmark.PageIndex + 1)
        Else
            Return False
        End If

        ' Save the modified document
        Dim addBookmarkPath As String = outputPath & "//AddBookmarkTest.pdf"
        If document.WriteToFilePath(addBookmarkPath) Then
            Console.WriteLine("Browse the changed file in " & addBookmarkPath)
            Return True
        Else
            Return False
        End If
    End Function

    ''' <summary>
    ''' Remove bookmark
    ''' </summary> 
    Private Function RemoveBookmark(document As CPDFDocument) As Boolean
        Dim bookmarkCount As Integer = document.GetBookmarkList().Count

        ' Remove the first bookmark
        document.RemoveBookmark(0)

        ' Check if the number of bookmarks has decreased by 1
        If bookmarkCount - document.GetBookmarkList().Count = 1 Then
            Console.WriteLine("Bookmark removed.")
        Else
            Return False
        End If

        ' Save the modified document
        Dim removeBookmarkPath As String = outputPath & "//RemoveBookmarkTest.pdf"
        If document.WriteToFilePath(removeBookmarkPath) Then
            Console.WriteLine("Browse the changed file in " & removeBookmarkPath)
            Return True
        Else
            Return False
        End If
    End Function

End Module

File: ComPDFKit for Windows/Examples/Samples/BookmarkTest/CS/BookmarkTest.cs
using ComPDFKit.PDFDocument;
using System;
using System.Collections.Generic;
using System.IO;

namespace BookmarkTest
{
    internal class BookmarkTest
    {
        static private string parentPath = Path.GetDirectoryName(Path.GetDirectoryName(Path.GetDirectoryName(System.IO.Directory.GetCurrentDirectory())));
        static private string outputPath = Path.Combine(parentPath, "Output", "CS");

        static void Main(string[] args)
        {
            #region Preparation work
            Console.WriteLine("Running Bookmark test sample…" + Environment.NewLine);
            SDKLicenseHelper.LicenseVerify();
            CPDFDocument document = CPDFDocument.InitWithFilePath("ThreeBookmark.pdf");

            if (!Directory.Exists(outputPath))
            {
                Directory.CreateDirectory(outputPath);
            }
            #endregion

            #region Sample 1: Access bookmark

            if (AccessBookmark(document))
            {
                Console.WriteLine("Check bookmark list done.");
            }
            else
            {
                Console.WriteLine("Check bookmark list failed.");
                Console.WriteLine("--------------------");
                Console.WriteLine("Stop.");
                Console.WriteLine("--------------------");
                Console.ReadLine();
                return;
            }
            document.Release();
            Console.WriteLine("--------------------");
             
            #endregion

            #region Sample 2: Create bookmark

            document = CPDFDocument.InitWithFilePath("ThreeBookmark.pdf");

            if (CreateBookmark(document))
            {
                Console.WriteLine("Add bookmark done.");
            }
            else
            {
                Console.WriteLine("Add bookmark failed.");
            }
            document.Release();
            Console.WriteLine("--------------------");

            #endregion

            #region Sample 3: Remove bookmark

            document = CPDFDocument.InitWithFilePath("ThreeBookmark.pdf");

            if (RemoveBookmark(document))
            {
                Console.WriteLine("Remove bookmark done.");
            }
            else
            {
                Console.WriteLine("Remove bookmark done.");
            }

            document.Release();
            Console.WriteLine("--------------------");

            #endregion
            
            Console.WriteLine("Done!");
            Console.WriteLine("--------------------");
            Console.ReadLine();
        }

        /// <summary>
        /// Access bookmark
        /// </summary> 
        static private bool AccessBookmark(CPDFDocument document)
        {
            List<CPDFBookmark> bookmarkList = document.GetBookmarkList();
            if (bookmarkList.Count == 3)
            {
                Console.WriteLine("Access bookmark list done.");
            }
            else
            {
                return false;
            }
            if (document.BookmarkForPageIndex(0).Title == "Bookmark1")
            {
                Console.WriteLine("Access bookmark for a page done.");
            }
            else
            {
                return false;
            }
            return true;
        }

        /// <summary>
        /// Create bookmark
        /// </summary> 
        static private bool CreateBookmark(CPDFDocument document)
        {
            var bookmarkCount = document.GetBookmarkList().Count;
            CPDFBookmark bookmark = new CPDFBookmark();
            bookmark.Title = "new bookmark";
            bookmark.PageIndex = 4;
            document.AddBookmark(bookmark);
            if (!(document.GetBookmarkList().Count - bookmarkCount == 1))
            {
                return false;
            }
            Console.WriteLine("Add bookmark in page {0}. ", bookmark.PageIndex + 1);
            string addBookmarkPath = Path.Combine(outputPath, "AddBookmarkTest.pdf");
            if (document.WriteToFilePath(addBookmarkPath))
            {
                Console.WriteLine("Browse the changed file in " + addBookmarkPath);
                return true;
            }
            else
            {
                return false;
            }
        }

        /// <summary>
        /// Remove bookmark
        /// </summary> 
        static private bool RemoveBookmark(CPDFDocument document)
        {
            var bookmarkCount = document.GetBookmarkList().Count;
            document.RemoveBookmark(0);
            if (!(bookmarkCount - document.GetBookmarkList().Count == 1))
            {
                return false;
            }
            string removeBookmarkPath = Path.Combine(outputPath, "RemoveBookmarkTest.pdf");
            if (document.WriteToFilePath(removeBookmarkPath))
            {
                Console.WriteLine("Browse the changed file in " + removeBookmarkPath);
                return true;
            }
            else
            {
                return false;
            }
        }
    }
}
